File: src/ur_four/scripts/waypoint_speed_control.cpp
#include <memory>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <moveit/robot_model_loader/robot_model_loader.h>
#include <moveit/robot_state/robot_state.h>
#include <rclcpp/rclcpp.hpp>
#include <trajectory_msgs/msg/joint_trajectory_point.hpp>
#include <Eigen/Geometry>



void EE_speed_control(const std::shared_ptr<rclcpp::Node> &node, moveit::core::RobotStatePtr &kinematic_state, moveit::planning_interface::MoveGroupInterface::Plan &plan, const std::string &end_effector, const double speed)
{
    // Get the number of waypoints and joint names
    int num_waypoints = plan.trajectory.joint_trajectory.points.size();
    const std::vector<std::string> joint_names = plan.trajectory.joint_trajectory.joint_names;

    // Set the joint positions of the first waypoint
    kinematic_state->setVariablePositions(joint_names, plan.trajectory.joint_trajectory.points[0].positions);

    Eigen::Affine3d current_end_effector_state = kinematic_state->getGlobalLinkTransform(end_effector);
    Eigen::Affine3d next_end_effector_state;
    double euclidean_distance, new_timestamp, old_timestamp, q1, q2, q3, dt1, dt2, v1, v2, a;
    trajectory_msgs::msg::JointTrajectoryPoint *prev_waypoint, *curr_waypoint, *next_waypoint;

    // Start iterating over the waypoints
    for (int i = 0; i < num_waypoints - 1; i++)
    {
        curr_waypoint = &plan.trajectory.joint_trajectory.points[i];
        next_waypoint = &plan.trajectory.joint_trajectory.points[i + 1];

        // Set joints for the next waypoint
        kinematic_state->setVariablePositions(joint_names, next_waypoint->positions);

        // Do forward kinematics to get cartesian positions of end effector for the next waypoint
        next_end_effector_state = kinematic_state->getGlobalLinkTransform(end_effector);

        // Get euclidean distance between the two waypoints
        euclidean_distance = (next_end_effector_state.translation() - current_end_effector_state.translation()).norm();

        // Updated for ROS 2 time handling
        new_timestamp = rclcpp::Duration(curr_waypoint->time_from_start).seconds() + (euclidean_distance / speed);
        old_timestamp = rclcpp::Duration(next_waypoint->time_from_start).seconds();

        // Update next waypoint timestamp if constraints allow
        if (new_timestamp > old_timestamp)
        {
            next_waypoint->time_from_start = rclcpp::Duration::from_seconds(new_timestamp);
        }
        else
        {
            // RCLCPP_WARN(rclcpp::get_logger("EE_speed_control"), "Average speed is too fast. Moving as fast as joint constraints allow.");
            next_waypoint->time_from_start = rclcpp::Duration::from_seconds(new_timestamp);
            RCLCPP_WARN(node->get_logger(), "%d -> %d :    New Timestamp: %f   |    Old Timestamp: %f", i, i + 1, new_timestamp, old_timestamp);
        }

        // Update the current_end_effector_state for the next iteration
        current_end_effector_state = next_end_effector_state;
    }

    // Update joint velocities and accelerations
    for (int i = 0; i < num_waypoints; i++)
    {
        curr_waypoint = &plan.trajectory.joint_trajectory.points[i];

        if (i > 0)
        {
            prev_waypoint = &plan.trajectory.joint_trajectory.points[i - 1];
        }
        if (i < num_waypoints - 1)
        {
            next_waypoint = &plan.trajectory.joint_trajectory.points[i + 1];
        }

        // Handle timestamp differences for ROS 2
        if (i == 0)
        {
            dt1 = dt2 = rclcpp::Duration(next_waypoint->time_from_start).seconds() - rclcpp::Duration(curr_waypoint->time_from_start).seconds();
        }
        else if (i < num_waypoints - 1)
        {
            dt1 = rclcpp::Duration(curr_waypoint->time_from_start).seconds() - rclcpp::Duration(prev_waypoint->time_from_start).seconds();
            dt2 = rclcpp::Duration(next_waypoint->time_from_start).seconds() - rclcpp::Duration(curr_waypoint->time_from_start).seconds();
        }
        else
        {
            dt1 = dt2 = rclcpp::Duration(curr_waypoint->time_from_start).seconds() - rclcpp::Duration(prev_waypoint->time_from_start).seconds();
        }

        // Iterate over all joints in waypoint
        for (size_t j = 0; j < joint_names.size(); j++)
        {
            if (i == 0)
            {
                q1 = next_waypoint->positions[j];
                q2 = curr_waypoint->positions[j];
                q3 = q1;
            }
            else if (i < num_waypoints - 1)
            {
                q1 = prev_waypoint->positions[j];
                q2 = curr_waypoint->positions[j];
                q3 = next_waypoint->positions[j];
            }
            else
            {
                q1 = prev_waypoint->positions[j];
                q2 = curr_waypoint->positions[j];
                q3 = q1;
            }

            if (dt1 == 0.0 || dt2 == 0.0)
            {
                v1 = v2 = a = 0.0;
            }
            else
            {
                v1 = (q2 - q1) / dt1;
                v2 = (q3 - q2) / dt2;
                a = 2.0 * (v2 - v1) / (dt1 + dt2);
            }

            // Set the velocity and acceleration
            curr_waypoint->velocities[j] = (v1 + v2) / 2.0;
            curr_waypoint->accelerations[j] = a;
        }
    }
}



int main(int argc, char * argv[])
{

  const std::string& ur1_ee = "UR1_wrist_3_link";
  const std::string& ur1_planning_group = "UR1_manipulator";

  // Initialize ROS and create the Node
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions node_options;
  node_options.automatically_declare_parameters_from_overrides(true);
  auto node = rclcpp::Node::make_shared("ORCA_MoveIt_Control", node_options);


  // We spin up a SingleThreadedExecutor for the current state monitor to get information
  // about the robot's state.
  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);
  std::thread([&executor]() { executor.spin(); }).detach();


  // Create a ROS logger
  auto const logger = rclcpp::get_logger("hello_moveit_UR1");

  // Create the MoveIt MoveGroup Interface
  using moveit::planning_interface::MoveGroupInterface;
  auto ur1_move_group_interface = MoveGroupInterface(node, ur1_planning_group);

  // Start the state monitor
  ur1_move_group_interface.startStateMonitor();

//   ur1_move_group_interface.setPlanningTime(10.0);  // Set a higher planning time, e.g., 10 seconds.
//   ur1_move_group_interface.setGoalTolerance(0.1);  // Set a reasonable goal tolerance.

  // Get and print the planning frame ID
  std::string planning_frame = ur1_move_group_interface.getPlanningFrame();
  RCLCPP_INFO(logger, "Planning frame: %s", planning_frame.c_str());

  // Wait for a while to ensure that the current state is properly updated
  rclcpp::sleep_for(std::chrono::seconds(2));

  // Get Current Robot EE Pose
  moveit::core::RobotStatePtr ur1_current_state = ur1_move_group_interface.getCurrentState();

  Eigen::Affine3d ur1_current_EE_state = ur1_current_state->getGlobalLinkTransform(ur1_ee);

  Eigen::Vector3d ur1_current_EE_position = ur1_current_EE_state.translation();

  RCLCPP_INFO(logger, "UR - 1 Current pose position: [x: %f, y: %f, z: %f]", 
    ur1_current_EE_position(0), 
    ur1_current_EE_position(1), 
    ur1_current_EE_position(2));
  
  //   std::vector<double> current_positions;
  //   moveit::core::RobotStatePtr current_state = ur1_move_group_interface.getCurrentState();
  //   current_state->copyJointGroupPositions("UR1_manipulator", current_positions);

  // Set a target Pose
  auto const ur1_target_pose = []{
    geometry_msgs::msg::Pose ur1_msg;
    ur1_msg.position.x = 0.73;
    ur1_msg.position.y = 0.3;
    ur1_msg.position.z = 1.05;

    // Define roll, pitch, and yaw
    double ur1_roll = 0.0; 
    double ur1_pitch = 3.14; 
    double ur1_yaw = 0.0; 

    // Convert roll, pitch, and yaw to quaternion
    tf2::Quaternion ur1_quaternion;
    ur1_quaternion.setRPY(ur1_roll, ur1_pitch, ur1_yaw);
    ur1_msg.orientation.x = ur1_quaternion.x();
    ur1_msg.orientation.y = ur1_quaternion.y();
    ur1_msg.orientation.z = ur1_quaternion.z();
    ur1_msg.orientation.w = ur1_quaternion.w();

    return ur1_msg;
  }();

  // Set the joint value target with approximation
  //   std::string end_effector_link = "UR1_wrist_3_link";  

  ur1_move_group_interface.setPoseTarget(ur1_target_pose);
  //   move_group_interface.setJointValueTarget(target_pose, end_effector_link);

  // Create a plan to that target pose
  auto const [ur1_success, ur1_plan] = [&ur1_move_group_interface]{
    moveit::planning_interface::MoveGroupInterface::Plan ur1_msg;
    auto const ur1_ok = static_cast<bool>(ur1_move_group_interface.plan(ur1_msg));
    return std::make_pair(ur1_ok, ur1_msg);
  }();

  if(ur1_success) {
    auto ur1_plan_copy = ur1_plan; // Copy the plan to avoid const issues
    EE_speed_control(node, ur1_current_state, ur1_plan_copy, ur1_ee, 0.4); // Pass the copy of the plan

    // Execute the plan after modifying it
    ur1_move_group_interface.execute(ur1_plan_copy);
    RCLCPP_INFO(logger, "UR - 1 has executed the planned motion.");


    // ur1_move_group_interface.asyncExecute(ur1_plan_copy);
    // RCLCPP_INFO(logger, "UR - 1 has executed the planned motion.");


  } else {
    RCLCPP_ERROR(logger, "UR - 1 Planning failed!");
  }

  // Wait for the motion to complete
//   rclcpp::sleep_for(std::chrono::seconds(2));

  ur1_current_state = ur1_move_group_interface.getCurrentState();

  ur1_current_state->update();

  ur1_current_EE_state = ur1_current_state->getGlobalLinkTransform(ur1_ee);

  ur1_current_EE_position = ur1_current_EE_state.translation();

  RCLCPP_INFO(logger, "UR - 1 New pose position: [x: %f, y: %f, z: %f]", 
    ur1_current_EE_position(0), 
    ur1_current_EE_position(1), 
    ur1_current_EE_position(2));

  // Keep the node alive while the motion is being executed
//   rclcpp::spin(node);

  // Shutdown ROS
  rclcpp::shutdown();
  return 0;
}

File: src/ur_four/scripts/waypoint_speed_control.py
#!/usr/bin/env python3

import os
import time
import numpy as np
import rclpy
from rclpy.logging import get_logger
from rclpy.duration import Duration
from moveit_configs_utils import MoveItConfigsBuilder
from ament_index_python import get_package_share_directory
from moveit.core.robot_model import RobotModel
from moveit.core.robot_state import RobotState
from moveit.planning import MoveItPy, MultiPipelinePlanRequestParameters
from moveit.core.kinematic_constraints import construct_joint_constraint
from geometry_msgs.msg import PoseStamped
import math
from pathlib import Path

from moveit.core.robot_trajectory import RobotTrajectory

def initialize_rclpy():
    rclpy.init()
    return get_logger("moveit_py.pose_goal")

def configure_moveit():

    ur_two = get_package_share_directory("ur_two")

    # Set workspace bounds (min and max corners of the planning volume)
    workspace_bounds = {
        "workspace_bounds": {
            "min_corner": [0.0, -0.85, 0.0],  # Set appropriate min corner values
            "max_corner": [1.50, 0.85, 2.0]     # Set appropriate max corner values
        }
    }

    moveit_configs = (
        MoveItConfigsBuilder(robot_name="ur", package_name="ur_two_moveit_config")
        .robot_description(file_path=os.path.join(ur_two, "urdf", "substitute.urdf"))
        .robot_description_semantic(Path("srdf") / "ur.srdf", {"name": "ur3e"})
        .moveit_cpp(file_path="config/moveit_cpp.yaml")
        .to_moveit_configs()
    )

    moveit_config_dict = moveit_configs.to_dict()

    # Update with workspace bounds
    moveit_config_dict.update(workspace_bounds)

    # Add QoS overrides for the /clock topic
    moveit_config_dict.update({
        'use_sim_time': True,
        'qos_overrides./clock.subscription.durability': 'volatile',  # 'transient_local'
        'qos_overrides./clock.subscription.history': 'keep_last',
        'qos_overrides./clock.subscription.depth': 100,
        'qos_overrides./clock.subscription.reliability': 'reliable',
    })

    return moveit_config_dict

def initialize_moveit(moveit_config_dict):
    ur_move = MoveItPy(node_name="moveit_py", config_dict=moveit_config_dict)
    logger.info("MoveItPy instance created")
    return ur_move

def quaternion_from_euler(roll, pitch, yaw):
    qx = np.sin(roll/2) * np.cos(pitch/2) * np.cos(yaw/2) - np.cos(roll/2) * np.sin(pitch/2) * np.sin(yaw/2)
    qy = np.cos(roll/2) * np.sin(pitch/2) * np.cos(yaw/2) + np.sin(roll/2) * np.cos(pitch/2) * np.sin(yaw/2)
    qz = np.cos(roll/2) * np.cos(pitch/2) * np.sin(yaw/2) - np.sin(roll/2) * np.sin(pitch/2) * np.cos(yaw/2)
    qw = np.cos(roll/2) * np.cos(pitch/2) * np.cos(yaw/2) + np.sin(roll/2) * np.sin(pitch/2) * np.sin(yaw/2)
    return [qx, qy, qz, qw]

def plan(planning_component,
    logger,
    single_plan_parameters=None,
    multi_plan_parameters=None,
):
    """Helper function to plan a motion."""
    # plan to goal
    logger.info("Planning trajectory")
    if multi_plan_parameters is not None:
        plan_result = planning_component.plan(
            multi_plan_parameters=multi_plan_parameters
        )
    elif single_plan_parameters is not None:
        plan_result = planning_component.plan(
            single_plan_parameters=single_plan_parameters
        )
    else:
        plan_result = planning_component.plan()

    return plan_result


def execute(robot, plan_result, controller_name, sleep_time=0.0):
    # execute the plan
    if plan_result:
        logger.info("Executing plan")
        robot_trajectory = plan_result.trajectory
        robot.execute(robot_trajectory, controllers=[controller_name])
    else:
        logger.error("Planning failed")

    time.sleep(sleep_time)



def EE_vel(robot_name, kinematic_state, plan_result, speed):

    if robot_name == "ur1":
        ee_link = "UR1_wrist_3_link"
        planner_group = "UR1_manipulator"

    elif robot_name == "ur2":
        ee_link = "UR2_wrist_3_link"
        planner_group = "UR2_manipulator"

    if plan_result:
        logger = get_logger("EE_vel")
        logger.info("Iterating through the waypoints")
        robot_trajectory = plan_result.trajectory

        # Use the get_robot_trajectory_msg method to get the trajectory message
        try:
            joint_trajectory_msg = robot_trajectory.get_robot_trajectory_msg().joint_trajectory
        except AttributeError:
            logger.error("The plan result does not have a valid joint_trajectory")
            return
        
        num_waypoints = len(joint_trajectory_msg.points)
        joint_names = joint_trajectory_msg.joint_names

        # Set the joint positions of the first waypoint
        joint_positions_init = np.array(joint_trajectory_msg.points[0].positions, dtype=np.float64)
        kinematic_state.set_joint_group_positions(planner_group, joint_positions_init)
        current_end_effector_state = kinematic_state.get_global_link_transform(ee_link).matrix()[:3, 3]

        for i in range(num_waypoints - 1):
            curr_waypoint = joint_trajectory_msg.points[i]
            next_waypoint = joint_trajectory_msg.points[i + 1]

            # Set joints for the next waypoint
            joint_positions = np.array(next_waypoint.positions, dtype=np.float64)
            kinematic_state.set_joint_group_positions(planner_group, joint_positions)

            # Do forward kinematics to get Cartesian positions of end effector for the next waypoint
            next_end_effector_state = kinematic_state.get_global_link_transform(ee_link).matrix()[:3, 3]

            # Get Euclidean distance between the two waypoints
            euclidean_distance = np.linalg.norm(np.array(next_end_effector_state) - np.array(current_end_effector_state))

            # Update timestamp
            new_timestamp = curr_waypoint.time_from_start.sec + curr_waypoint.time_from_start.nanosec * 1e-9 + (euclidean_distance / speed)
            old_timestamp = next_waypoint.time_from_start.sec + next_waypoint.time_from_start.nanosec * 1e-9

            # Update next waypoint timestamp if constraints allow
            if new_timestamp > old_timestamp:
                next_waypoint.time_from_start = Duration(seconds=int(new_timestamp), nanoseconds=int((new_timestamp % 1) * 1e9))
            else:
                logger.warn(f"{i} -> {i+1} :    New Timestamp: {new_timestamp}   |    Old Timestamp: {old_timestamp}")

            # Update the current_end_effector_state for the next iteration
            current_end_effector_state = next_end_effector_state

        # Update joint velocities and accelerations
        for i in range(num_waypoints):
            curr_waypoint = joint_trajectory_msg.points[i]

            if i > 0:
                prev_waypoint = joint_trajectory_msg.points[i - 1]
            if i < num_waypoints - 1:
                next_waypoint = joint_trajectory_msg.points[i + 1]

            # Handle timestamp differences
            if i == 0:
                dt1 = dt2 = (next_waypoint.time_from_start.sec + next_waypoint.time_from_start.nanosec * 1e-9) - (curr_waypoint.time_from_start.sec + curr_waypoint.time_from_start.nanosec * 1e-9)
            elif i < num_waypoints - 1:
                dt1 = (curr_waypoint.time_from_start.sec + curr_waypoint.time_from_start.nanosec * 1e-9) - (prev_waypoint.time_from_start.sec + prev_waypoint.time_from_start.nanosec * 1e-9)
                dt2 = (next_waypoint.time_from_start.sec + next_waypoint.time_from_start.nanosec * 1e-9) - (curr_waypoint.time_from_start.sec + curr_waypoint.time_from_start.nanosec * 1e-9)
            else:
                dt1 = dt2 = (curr_waypoint.time_from_start.sec + curr_waypoint.time_from_start.nanosec * 1e-9) - (prev_waypoint.time_from_start.sec + prev_waypoint.time_from_start.nanosec * 1e-9)

            # Iterate over all joints in waypoint
            for j in range(len(joint_names)):
                if i == 0:
                    q1 = next_waypoint.positions[j]
                    q2 = curr_waypoint.positions[j]
                    q3 = q1
                elif i < num_waypoints - 1:
                    q1 = prev_waypoint.positions[j]
                    q2 = curr_waypoint.positions[j]
                    q3 = next_waypoint.positions[j]
                else:
                    q1 = prev_waypoint.positions[j]
                    q2 = curr_waypoint.positions[j]
                    q3 = q1

                if dt1 == 0.0 or dt2 == 0.0:
                    v1 = v2 = a = 0.0
                else:
                    v1 = (q2 - q1) / dt1
                    v2 = (q3 - q2) / dt2
                    a = 2.0 * (v2 - v1) / (dt1 + dt2)

                # Set the velocity and acceleration
                curr_waypoint.velocities[j] = (v1 + v2) / 2.0
                curr_waypoint.accelerations[j] = a

        return plan_result



def main():
    global logger
    logger = initialize_rclpy()

    ur1_ee_link = "UR1_wrist_3_link"
    ur2_ee_link = "UR2_wrist_3_link"

    moveit_config = configure_moveit()
    ur_move = initialize_moveit(moveit_config)
    
    ur1 = ur_move.get_planning_component("UR1_manipulator")
    # ur2 = ur_move.get_planning_component("UR2_manipulator")

    

    ur_model = ur_move.get_robot_model()
    ur_state = RobotState(ur_model)

    # set plan start state to current state
    ur1.set_start_state_to_current_state()

    ur1_roll = 0.0
    ur1_pitch = 1.57
    ur1_yaw = 0.0

    ur1_ort = quaternion_from_euler(ur1_roll, ur1_pitch, ur1_yaw)

    ur1_pose_goal = PoseStamped()
    ur1_pose_goal.header.frame_id = "world"
    ur1_pose_goal.pose.orientation.x = ur1_ort[0]
    ur1_pose_goal.pose.orientation.y = ur1_ort[1]
    ur1_pose_goal.pose.orientation.z = ur1_ort[2]
    ur1_pose_goal.pose.orientation.w = ur1_ort[3]
    ur1_pose_goal.pose.position.x = 0.75
    ur1_pose_goal.pose.position.y = 0.0    # 0.25 & 0.05
    ur1_pose_goal.pose.position.z = 1.0
    ur1.set_goal_state(pose_stamped_msg=ur1_pose_goal, pose_link="UR1_wrist_3_link")

    # plan to goal
    ur1_planned_trajectory = plan(ur1, logger)

    # ur1_fixed_vel_trajectory = EE_vel("ur1", ur_state, ur1_planned_trajectory, 0.0001)

    # execute(ur_move, ur1_fixed_vel_trajectory)

    ur1_controller_name = "joint_trajectory_controller_ur1";

    execute(ur_move, ur1_planned_trajectory, ur1_controller_name)

if __name__ == "__main__":
    main()